pallets/amm/src/pool_structs.rs

```diff
@@ -0,0 +1,180 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Groups common pool related structures
+
+use parallel_primitives::{Balance, CurrencyId, Rate};
+use sp_runtime::{traits::Saturating, ArithmeticError, DispatchError, FixedPointNumber};
+
+// Amplification Coefficient Weight.
+//
+// In this pallet, the actual amplification coefficient will be `exchange_rate` * `ACW`.
+const ACW: Rate = Rate::from_inner(Rate::DIV / 100 * 50); // 50%
+
+#[derive(
+    Clone,
+    PartialEq,
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    sp_runtime::RuntimeDebug,
+)]
+pub enum SwapType {
+    Buy,
+    Sell,
+}
+
+#[derive(Debug)]
+// Wrapper around the result of `Pallet::calculate_amount`
+pub struct AmountEvaluation {
+    pub account_amount: Balance,
+    pub pool_amount: Balance,
+}
+
+#[derive(
+    Clone,
+    PartialEq,
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    sp_runtime::RuntimeDebug,
+)]
+pub struct LiquidityProviderAmounts {
+    pub base_amount: Balance,
+    pub quote_amount: Balance,
+}
+
+#[derive(
+    Clone,
+    PartialEq,
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    sp_runtime::RuntimeDebug,
+)]
+pub struct Pool {
+    pub base_amount: Balance,
+    pub quote_amount: Balance,
+    pub base_asset: CurrencyId,
+    pub quote_asset: CurrencyId,
+}
+
+pub struct StandardSwap;
+
+pub struct StableSwap;
+
+pub struct StakingSwap;
+
+pub trait StabilityPool {}
+
+impl StabilityPool for () {}
+
+pub trait AMMCurve {
+    // Calculates the amount according to the underlying formula and the provided pool.
+    fn calculate_amount(
+        exchange_rate: Rate,
+        new_amount: Balance,
+        pool: &Pool,
+    ) -> Result<Balance, DispatchError>;
+}
+
+impl AMMCurve for StandardSwap {
+    // let [x|y] = k / [x|y];
+    fn calculate_amount(
+        _: Rate,
+        new_amount: Balance,
+        pool: &Pool,
+    ) -> Result<Balance, DispatchError> {
+        let k = pool
+            .base_amount
+            .checked_mul(pool.quote_amount)
+            .ok_or_else::<DispatchError, _>(|| ArithmeticError::Overflow.into())?;
+        let result = k
+            .checked_div(new_amount)
+            .ok_or_else::<DispatchError, _>(|| ArithmeticError::DivisionByZero.into())?;
+        Ok(result)
+    }
+}
+
+impl AMMCurve for StableSwap {
+    // let [x|y] = (k * (4*A*k + k - 4*A*[x|y])) / (4 * (A*k + [x|y]));
+    fn calculate_amount(
+        exchange_rate: Rate,
+        new_amount: Balance,
+        pool: &Pool,
+    ) -> Result<Balance, DispatchError> {
+        let k = pool
+            .base_amount
+            .checked_add(pool.quote_amount)
+            .ok_or_else::<DispatchError, _>(|| ArithmeticError::Overflow.into())?;
+        let evaluation_option = || {
+            let a_multiplied_by_k = amplification_coeficient_mul(exchange_rate, k)?;
+            let _4_multiplied_by_a_multiplied_by_x =
```

Comment:
@c410-f3r I just noticed this. Can we remove `_` at the beginning of variable name? Usually we do this when variable is not in use.

Reply:
Sure